new/rfcXXX-new-attributes-api.md

```diff
@@ -0,0 +1,151 @@
+Status: Draft
+Type: Standards Track
+---
+
+# New Attributes API
+
+This RFC proposes a new API for setting and accessing attributes in
+cookbooks. The goals of this new API are:
+
+* Separate desired and observed state.
+* Structure the syntax in a way that "trainwreck" errors
+  (`NoMethodError` for `[]` on `nil`) can be avoided.
+* Structure the syntax in a way that attribute tracing can be
+  implemented simply.
+* Separate node-specific persistent storage from
+  cookbook/role/environment attributes.
+* Structure the syntax in a way that can be implemented efficiently.
+
+The new API is added alongside the existing API, so existing code will
+work without modification.
+
+## Motivation
+
+There are several motivations for this change.
+
+### Separate Desired and Observed State
+
+In the current attributes system, system "facts" determined by Ohai are
+intermixed with user attributes, which are generally used to define the
+desired state of the system. For example, `node[:fqdn]` gives the
+system's FQDN, as determined by Ohai at the beginning of the Chef run,
+and identical syntax is used to reference a user-defined attribute, such
+as `node[:mysql][:version]`. As a result, it's not possible to know
+which attributes are "facts" except by committing to memory the list of
+attributes provided by Ohai. Defining a new syntax where Ohai and
+user-defined attributes are accessed by different methods makes cookbook
+code clearer.
+
+Additionally, separating Ohai data from user-defined attributes
+eliminates the possibility of attribute name collisions. For example, a
+user might wish to set the FQDN of a host with Chef and use the `fqdn`
+attribute to define the desired FQDN, or create an Ohai plugin to detect
+the currently installed version of MySQL, storing this as
+`node[:mysql][:version]`. With the current attributes system, the user
+can't use these attribute names because they conflict with existing Ohai
+or cookbook attributes.
+
+### Eliminate Trainwreck Errors
+
+The current attribute system presents node data as a nested collection
+of Hash-like objects. There are some unfortunate limitations of this
+model that lead to frustrating behavior for users. In particular, code
+like this:
+
+```ruby
+node[:foo][:bar][:baz]
+```
+
+May raise a `NoMethodError` with the message:
+
+```
+undefined method `[]' for nil:NilClass
+```
+
+Note that in this example, it's ambiguous as to which part of the
+attribute structure is missing, as this error will occur in an identical
+way if either `node[:foo]` or `node[:foo][:bar]` are `nil`.
+
+The UX of this error case cannot be significantly improved without
+abandoning the nested Hash model. If an object other than `nil` was
+returned for accessing a missing key, then code like
+`do_something if node[:some_attribute_present]` would break (Ruby does
+not allow custom false-y types), and monkey patching `nil` is
+error-prone because `nil` is an immediate object (there is only one
+instance).
```

Comment:
I didn't want to clutter this section too much, but I guess that kind of thing is what the rationale section is for.